File: Structural/Proxy/CSharp/VideoProxy.cs
// VideoProxy.cs

using System;
using System.Collections.Generic;

public class VideoProxy : IVideoService {
  private readonly RealVideoService realVideoService = new RealVideoService();
  private readonly Dictionary<string, string> cache = new Dictionary<string, string>();

  public void LoadVideo(string videoName) {
    if (cache.ContainsKey(videoName)) {
      Console.WriteLine($"Retrieving cached video: {videoName}");
      Console.WriteLine($"Playing video: {videoName}");
    } else {
      realVideoService.LoadVideo(videoName);
      cache[videoName] = "Cached";
    }
  }
}

File: Structural/Proxy/Cpp/VideoProxy.cpp
// VideoProxy.cpp

#include "VideoProxy.h"
#include <iostream>

void VideoProxy::loadVideo(const std::string& videoName) {
  if (cache.find(videoName) != cache.end()) {
    std::cout << "Retrieving cached video: " << videoName << std::endl;
    std::cout << "Playing video: " << videoName << std::endl;
  } else {
    realVideoService.loadVideo(videoName);
    cache[videoName] = true;
  }
}
